팀 배정 / 팀 정보 조회 기능 개발

1seyoung opened the pull request:
### 구현 내용

팀 자동 배정 및 수동 배정 API를 구현하고 테스트를 진행함.
현재 팀 자동 배정 및 전체 조회는 테스트 완료, 팀 수동 배정 및 개별 조회는 진행 예정.

**팀 자동 배정**
- 학생들을 균등하게 팀에 배정하는 로직을 구현
- 배정된 팀 정보를 데이터베이스에 저장 후 검증
- 최초 1회 진행

**팀 수동 배정**
- 자동 배정 이후 사용 가능
- 특정 팀을 선택해서 학생 정보 수정

**기능 테스트**
- [x] 팀 자동 배정 테스트
- [ ] 팀 수동 배정 테스트(진행 예정)
- [x] 팀 전체 조회 테스트
- [ ] 팀 개별 조회 테스트 (진행 예정)

### 고민 포인트, 리뷰 시 참고 사항(코드)
팀 수동 배정 및 개별 조회에 대한 테스트와 오류 개선 진행 중
Mapper 코드가 지저분하고 중복이 많음
> SQL 쿼리와 로직이 반복되는 부분이 많아 리팩토링 필요
> 여러 개의 JOIN을 사용하는 복잡한 쿼리가 많아 가독성이 떨어짐
> 현재 중복된 SQL 및 비효율적인 쿼리 사용 여부 검토필요


### 메인 리뷰어 지정 및 Due Date
@blue000927 

### 관련 이슈 
#15 
#16 
#17 

### 체크리스트
- [x]  PR 제목을 명령형으로 작성했습니다.
- [x]  PR을 연관되는 github issue에 연결했습니다.
- [x]  리뷰 리퀘스트 전에 셀프 리뷰를 진행했습니다.

1seyoung committed:
feat : 팀 배정 로직 구현 , 팀 정보 조회 API
- 자동 팀 배정
- 수동 팀 배정
- 팀 정보 조회

1seyoung committed:
bugfix: 팀 정보 조회 로직 수정
- studentid / member id 혼동 문제
- 팀 배정 오류 해결

blue000927 commented on src/main/java/com/gatheria/controller/TeamController.java:
```diff
@@ -0,0 +1,71 @@
+    this.teamService = teamService;
+  }
+  
+  @PostMapping("/auto-assign")
+  public ResponseEntity<Void> autoAssignTeams(
+      @PathVariable Long lectureId,
+      @RequestParam int teamSize,
+      @RequestAttribute("authInfo") AuthInfo authInfo) {
+
+    teamService.autoAssignTeams(lectureId, teamSize, authInfo);
+    return ResponseEntity.ok().build();
+  }
+
```
질문) 자동 배정된 결과를 프론트에서 알 필요는 없는 건가요?

1seyoung replied:
자동 배정이 되면 프론트에서 팀 리스트를 다시 가져오는 API를 호출하도록 설정하는 게 좋을 것 같아서, 처음에는 완료만 되면 충분하다고 생각했는데, 그래도 다시 호출되는 게 나을까요?

pjy1368 replied:
아하 그런 메커니즘이면 유지해도 좋습니다. 고치라는 이야기는 아니에요~

blue000927 commented on src/main/java/com/gatheria/controller/TeamController.java:
```diff
@@ -0,0 +1,71 @@
+  ) {
+    teamService.manualAssignTeams(lectureId, teamId, studentId, authInfo);
+    return ResponseEntity.ok().build();
+  }
```
ditto (결과물 반환 관련 질문)

blue000927 commented on src/main/java/com/gatheria/controller/TeamController.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  @GetMapping("/{teamId}")
+  public ResponseEntity<TeamResponseDto> showTeam(
```
위는 find고 아래는 show인데 조회 메서드 네이밍을 통일하면 어떨까요? (ex. show)

1seyoung replied:
넵 수정하겠습니다

blue000927 commented on src/main/java/com/gatheria/domain/Team.java:
```diff
@@ -0,0 +1,28 @@
+  private Long id;
+  private String name;
+  private Long lectureId;
+  private LocalDateTime createdAt;
+  private LocalDateTime updatedAt;
```
생성일, 수정일 필드는 모든 엔티티가 공통적으로 붙는 것 같은데 별도 클래스로 빼 볼 수 없을까요?
혹은 BaseEntity 라는 abstract class를 정의하여 상속하는 것도 방법일듯하네요 :)

1seyoung replied:
전체 엔티티 코드들 모두 BaseEntity를 상속받는 방향으로 수정하면 될까요?

pjy1368 replied:
네네

1seyoung replied:
id, createdAt, updatedAt만 따로 BaseEntity로 구현했고, Member, Student, Instructor 도메인을 제외하고 상속하도록 수정했습니다.  
Member는 abstract 클래스라 BaseEntity를 상속받는 게 맞는지 확신이 안 들어서 일단 제외해두었습니다

blue000927 replied:
멤버는 중복되겠지만 createdAt, updatedAt을 한 번 더 적어주어야겠군요. (다중 상속 문제)

blue000927 commented on src/main/java/com/gatheria/domain/TeamMember.java:
```diff
@@ -0,0 +1,35 @@
+  public void updateTeamId(Long newTeamId) {
+    this.teamId = newTeamId;
+    this.updatedAt = LocalDateTime.now();
+  }
```
updatedAt은 자동으로 데이터베이스 단에서 update될 때 updatedAt을 수정해 주는 방법이 있을거예요!
(한 번 찾아보시죠 ㅎㅎ)

1seyoung replied:
요 부분도 BaseEntity 로 변경하면서 제거했습니다

blue000927 commented on src/main/java/com/gatheria/mapper/LectureMapper.java:
```diff
@@ -17,8 +18,11 @@ public interface LectureMapper {
 
-  @Select("SELECT * FROM lectures WHERE id IN (${lectureIds})")
-  List<Lecture> findLecturesByIds(@Param("lectureIds") List<Long> lectureIds);
+  //여기 수정 중
+  List<Lecture> findLecturesByIds(@Param("list") List<Long> lectureIds);
```
어떤 식으로 수정하실거죠?

1seyoung replied:
이미 오류 잡으면서 로그 찍어보던 곳인데 주석 지우는 것을 까먹은 것 같습니다. 수정해두겠습니다~

blue000927 commented on src/main/java/com/gatheria/service/TeamService.java:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    teamMapper.saveTeamMembers(teamMembers);
+  }
```
객체지향 원칙 중 캡슐화를 적용해볼 방법이 없을까요?
적절히 우리가 정의한 객체 내부 메서드를 정의해서 지금 절차지향적으로 작성한 코드를 객체 간의 협력으로 고칠 수 있어 보여요!

blue000927 replied:
그 외 아래 코드도 살펴보면서 TeamService 전체적으로 리팩토링 고민해 보시면 좋겠습니다.

blue000927 commented on src/main/java/com/gatheria/service/TeamService.java:
```diff
@@ -0,0 +1,233 @@
+        .toList();
+
+    return TeamResponseDto.of(team.getId(), team.getName(), lectureId, studentResponses);
+  }
```
예를 들면 DTO를 만들어내는 방법은 여기서 정의하지 않고, DTO가 엔티티를 파라미터로 받아서 generate할수도있습니다.

blue000927 commented on src/main/java/com/gatheria/service/TeamService.java:
```diff
@@ -0,0 +1,233 @@
+      throw new IllegalArgumentException("Invalid Lecture");
+    }
+
+    if (authInfo.getRole() == MemberRole.INSTRUCTOR) {
+      if (!Objects.equals(lecture.getInstructorId(), authInfo.getMemberId())) {
+        throw new IllegalArgumentException("교수자는 해당 수업에 접근 권한 x");
+      }
+    } else if (authInfo.getRole() == MemberRole.STUDENT) {
+      boolean isEnrolled = lectureMapper.existEnrollmentByStudentIdAndLectureID(
+          authInfo.getMemberId(), lectureId);
+      if (!isEnrolled) {
+        throw new IllegalArgumentException("학생이 해당 수업에 등록 x");
+      }
+
+    } else {
+      throw new IllegalArgumentException("Invalid Role");
+    }
```
LectureService에서도 이 코드가 있었는데 어떻게 중복을 피할 수 있을까요?

1seyoung replied:
Util 클래스를 만드는 게 좋을까요, 아니면 인터페이스를 사용해서 두 서비스 코드에서 모두 활용할 수 있도록 하는 게 나을까요?

1seyoung committed:
Merge branch 'feature/lecture' into feature/create-team

1seyoung committed:
chore : authinfo 등 수정 사항 반영

1seyoung committed:
Merge main into feature/create-team

1seyoung committed:
feat : BaseEntity 추상 클래스 추가 및 기존 엔티티 클래스 리팩토링
- Team, LectureStudent, TeamMember, Lecture 등의 엔티티 클래스가 BaseEntity를 상속받도록 수정
- updatedAt 수동 업데이트 코드 제거

1seyoung committed:
Merge branch 'main' into feature/create-team

1seyoung committed:
feat: 수업 내 팀원 정보 조회 기능 구현 및 팀 기능 리팩토링

blue000927 commented on src/main/java/com/gatheria/domain/Team.java:
```diff
@@ -0,0 +1,38 @@
+
+  public static List<Team> createTeams(Long lectureId, int teamCount) {
+    List<Team> teams = new ArrayList<>();
+    for (int i = 0; i < teamCount; i++) {
+      String teamName = String.valueOf('A' + i);
+      teams.add(Team.of(teamName, lectureId));
+    }
```
질문) 팀 이름은 이렇게 'A1', 'A2', ... 로만 만드시는 것 의도하신건가요?

1seyoung replied:
팀이 생성될 때 팀 이름이 비워져있는게 싫어서 해당 로직 추가해두었습니다. 나중에 필요하면 학생들이 수정하는 API 추가하면 될 것 같아서 유지했습니다.

blue000927 commented on src/main/java/com/gatheria/domain/TeamMember.java:
```diff
@@ -0,0 +1,60 @@
+    this.teamId = newTeamId;
+  }
+
+  public static List<TeamMember> createTeamMembers(List<Team> teams, List<Long> studentIds,
+      Long lectureId, int teamSize) {
```
TeamMember는 단수 객체인데, 단수 객체가 복수 객체인 List를 반환하는 것은 어색한 것 같아요. TeamMembers 라는 클래스를 정의하고 그 클래스가 `List<TeamMember>`를 갖도록 구성해 보시면 어떨까요?

참고) 일급 컬렉션 - https://jojoldu.tistory.com/412

1seyoung replied:
TeamMembers 일급 컬렉션을 구현하여 기존 TeamMember.createTeamMembers()를 대체하는 것까지 완료했습니다! 이 컬렉션을 팀 서비스의 다른 부분에서도 활용할한 곳이 있을까요? 팀 조회 이런 부분에서도 활용하게 수정하는게 좋을지 궁금합니다~!!

blue000927 replied:
DB로직과 섞이다보니 크게 보이진 않지만 나중에 얼마든지 사용 가능할거예요~

blue000927 commented on src/main/java/com/gatheria/dto/response/TeamResponseDto.java:
```diff
@@ -0,0 +1,32 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+/**
+ * List<TeamResponseDto>  // ← 여러 팀 └── List<StudentInfoResponseDto>  // ← 팀 안의 학생들
+ */
```
주석 없어도 되지 않을까요?

1seyoung replied:
넵 다시 설계하다가 추가한건데 까먹었습니다~~ 삭제하겠습니다

blue000927 commented on src/main/java/com/gatheria/mapper/MemberMapper.java:
```diff
@@ -28,4 +32,10 @@ public interface MemberMapper {
   Student findStudentById(Long studentId);
+
+
+  //---TeamService.java ---
+  @MapKey("studentId")
```
이건 무슨 의미인가요?

1seyoung replied:
@MapKey는 MyBatis에서 결과를 Map으로 반환할 때,
그 Map의 key로 사용할 필드를 지정하는 어노테이션입니다.

Map<Long, StudentInfoResponseDto> studentDtoMap = memberMapper.findStudentInfoMapByIds(studentIds); // TeamService 104line

코드에서 Map의 key로 studentId를 사용하고 싶어서
@MapKey("studentId") 어노테이션을 추가했습니다.

이 외에도 다른 방법이 있을까요?

blue000927 replied:
아뇽 질문이었고 필요한 문법으로 보이네요!

blue000927 commented on src/main/java/com/gatheria/service/LectureService.java:
```diff
@@ -27,9 +26,10 @@ public LectureService(LectureMapper lectureMapper, MemberMapper memberMapper) {
-      throw new RuntimeException();
+    if (!authInfo.isInstructor()) {
+      throw new RuntimeException("교수자 권한이 필요");
     }
```
authInfo 안에 throw new RuntimeException()을 던지는 validateInstructor()? 같은 메서드를 두어도 좋을거같아요.

blue000927 commented on src/main/java/com/gatheria/service/TeamService.java:
```diff
@@ -0,0 +1,252 @@
+      teamMapper.saveTeamMember(newMember);
+    } else {
+      existingMember.updateTeamId(targetTeam.getId());
+      teamMapper.updateTeamMember(existingMember);
```
"upsert" 연산에 대해 학습해 보고 적용해 보면 어떨까요?

1seyoung committed:
feat: TeamMembers 일급 컬렉션 구현 및 적용 + 주석 삭제
- TeamMember 리스트를 관리하는 TeamMembers 일급 컬렉션 클래스 구현
- autoAssignTeams 메서드에서 TeamMembers 사용하도록 수정
- 팀원 생성 로직을 TeamMember에서 TeamMembers로 이동

1seyoung committed:
refactor : AuthInfo 클래스에 권한 검증 메서드 추가 및 권한 로직 개선
- AuthInfo 클래스에 validateInstructor(), validateStudent() 메서드 추가
- 연관 로직 수정

blue000927 commented on src/main/java/com/gatheria/mapper/MemberMapper.java:
```diff
@@ -28,4 +32,8 @@ public interface MemberMapper {
   @Select("SELECT * FROM students WHERE id = #{studentId}")
   Student findStudentById(Long studentId);
+
+  @MapKey("studentId")
+  Map<Long, StudentInfoResponseDto> findStudentInfoMapByIds(
+      @Param("studentIds") List<Long> studentIds);
```
controller에서 쓰는 dto와 mapper에서 쓰는 dto가 혼용되어 헷갈리는 것 같아요.
패키지라도 분리해보면 어떨까요?

blue000927 commented on src/main/resources/mapper/MemberMapper.xml:
```diff
@@ -1,70 +1,108 @@
+  </update>
+
+
+  <select id="findStudentInfoMapByIds" resultType="com.gatheria.dto.StudentInfoResponseDto">
+    SELECT
+    s.id AS studentId,
+    m.name,
+    m.email
+    FROM students s
+    JOIN members m ON s.member_id = m.id
+    WHERE s.id IN
+    <foreach collection="studentIds" item="id" open="(" separator="," close=")">
+      #{id}
+    </foreach>
+  </select>
+
+
+  <select id="findStudentInfoMapByIds"
+    resultType="com.gatheria.dto.response.StudentInfoResponseDto">
+    SELECT
+    s.id AS studentId,
+    m.name,
+    m.email
+    FROM students s
+    JOIN members m ON s.member_id = m.id
+    WHERE s.id IN
+    <foreach collection="studentIds" item="id" open="(" separator="," close=")">
+      #{id}
+    </foreach>
+  </select>
```
이런 부분이 참 고민이되네요.
일반적으로 서비스 회사 기준으로 join을 하기보다는 쿼리를 2개로 쪼갠 뒤 서비스 레이어에서 단일 select 쿼리 2번 호출하여 결과물을 조합하는 경우가 많습니다. (만약 그렇게한다면 별도 Mapper를 위한 DTO는 없어져도 됩니다.)

하지만 지금 단계에서 이것을 모두 지키기는 어려울 수 있으니, 이번 PR에서는 작업하지 않고 향후 고려해 주시면 감사하겠습니다 :)

blue000927 replied:
이렇게 하는 이유는 DB 쿼리를 단순하게 가져가야 유지 보수하기 좋기 때문입니다.
우리가 자바 코드를 유지 보수하는 일은 객체 지향 기법 등을 사용해서 얼마든지 개선할 수 있는데, 쿼리가 복잡해져버리면 조금만 요구 사항이 달라져도 쿼리 자체를 뜯어 고치기 쉽지가 않습니다..!

1seyoung replied:
넵 해당 부분은 이슈에 적어두고 다른 기능 완성 후에 다시 체크해보겠습니다
